samples/Common.c

```diff
@@ -704,24 +715,70 @@ VOID sampleBandwidthEstimationHandler(UINT64 customData, DOUBLE maximumBitrate)
 VOID sampleSenderBandwidthEstimationHandler(UINT64 customData, UINT32 txBytes, UINT32 rxBytes, UINT32 txPacketsCnt, UINT32 rxPacketsCnt,
                                             UINT64 duration)
 {
-    UNUSED_PARAM(customData);
     UNUSED_PARAM(duration);
-    UNUSED_PARAM(rxBytes);
-    UNUSED_PARAM(txBytes);
+    UINT64 videoBitrate, audioBitrate;
+    BOOL locked = FALSE;
     UINT32 lostPacketsCnt = txPacketsCnt - rxPacketsCnt;
-    UINT32 percentLost = lostPacketsCnt * 100 / txPacketsCnt;
-    UINT32 bitrate = 1024;
-    if (percentLost < 2) {
-        // increase encoder bitrate by 2 percent
-        bitrate *= 1.02f;
-    } else if (percentLost > 5) {
-        // decrease encoder bitrate by packet loss percent
-        bitrate *= (1.0f - percentLost / 100.0f);
-    }
-    // otherwise keep bitrate the same
-
-    DLOGV("received sender bitrate estimation: suggested bitrate %u sent: %u bytes %u packets received: %u bytes %u packets in %lu msec", bitrate,
-          txBytes, txPacketsCnt, rxBytes, rxPacketsCnt, duration / 10000ULL);
+    UINT8 percentLost = (txPacketsCnt > 0) ? (lostPacketsCnt * 100 / txPacketsCnt) : 0;
+    SampleStreamingSession* pSampleStreamingSession = (SampleStreamingSession*) customData;
+    if (pSampleStreamingSession == NULL) {
+        DLOGW("Invalid streaming session (NULL object)");
+        return;
+    }
+
+    // Calculate packet loss
+    DOUBLE currentPacketLoss = (DOUBLE) percentLost;
+    pSampleStreamingSession->twccMetadata.averagePacketLoss =
+        EMA_ACCUMULATOR_GET_NEXT(pSampleStreamingSession->twccMetadata.averagePacketLoss, currentPacketLoss);
+
+    UINT64 currentTimeMs = GETTIME();
+    UINT64 timeDiff = currentTimeMs - pSampleStreamingSession->twccMetadata.lastAdjustmentTimeMs;
+    if (timeDiff < ADJUSTMENT_INTERVAL_SECONDS) {
+        // Too soon for another adjustment
+        return;
+    }
+
+    MUTEX_LOCK(pSampleStreamingSession->twccMetadata.updateLock);
+    locked = TRUE;
+    videoBitrate = pSampleStreamingSession->twccMetadata.currentVideoBitrate;
+    audioBitrate = pSampleStreamingSession->twccMetadata.currentAudioBitrate;
+    if (locked) {
+        MUTEX_UNLOCK(pSampleStreamingSession->twccMetadata.updateLock);
+        locked = FALSE;
+    }
+
+    if (pSampleStreamingSession->twccMetadata.averagePacketLoss <= 5) {
+        // increase encoder bitrate by 5 percent with a cap at MAX_BITRATE
```

Comment:
Is this correlation of avg packet loss with bitrate a recommendation from a spec? How did we deduced this calculation?

Reply:
There is no spec recommendation on how the values should be set, neither is there any spec on how the algorithm should be implemented. This is an adaptation of the google congestion loss based algorithm. I set it to 5% to allow a decent increase with good bitrate. This callback can be modified to change these values based on customer set up and tolerances. It is an example implementation.

Reply:
One thing we've missed on when implementing features like this that have possibly multiple toggles instead of exposing all the toggles to the customer (which we do some times) OR hard coding those values (which we also do sometimes) is to instead have like 2-3 profiles / strategies.  


Examining this whole feature at a high level I'm thinking if instead of even exposing TWCC as an option would it better customer experience to instead expose options like:

`AGGRESSIVE_PACKET_LOSS`, `MODERATE_PACKET_LOSS`, `LOW_PACKET_LOSS`

Perhaps in these options the first two we enable TWCC and we set different values for average packet loss threshold and for low packet loss we just don't do TWCC at all? Basically I'm thinking about a higher level profile/mode the SDK runs in and we hide the fine grained control of this packet loss behind that?

Reply:
Since this PR is for the sample illustration of the feature and what I'm talking about is core to the SDK itself, we can consider that as a future enhancement.